llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -8663,6 +8663,81 @@ VPReplicateRecipe *VPRecipeBuilder::handleReplication(Instruction *I,
   return Recipe;
 }
 
+/// Examines reduction operations to see if the target can use a cheaper
+/// operation with a wider per-iteration input VF and narrower PHI VF.
+/// Returns a struct containing the ratio between the two VFs and other cached
+/// information, or null if no scalable reduction was found.
+static std::optional<PartialReductionChain>
+getScaledReduction(PHINode *PHI, const RecurrenceDescriptor &Rdx,
```

Comment:
Could be moved to VPRecipeBuilder, thus not needing TTI/CM passed?

Reply:
Nice idea, done.